File: cpp/tests/test_ara_star.cpp
#include <memory>
#include <sstream>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/search/ara_star.hpp"
#include "navigation/global_planning/search/astar.hpp"
#include "navigation/maps/loader.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Cell;

namespace {

core::ParamSet cfg(const std::string& algo) {
  return core::ParamSet::from_yaml(test::repo_path("configs/global_planning/" + algo + ".yaml"));
}

// Writes an ARA* config with the given schedule so tests can force a specific
// eps_start -> eps_final sweep without depending on the shipped defaults.
core::ParamSet ara_cfg(double eps_start, double eps_final, double eps_step) {
  std::ostringstream y;
  y << "algorithm: ara_star\ncategory: global_planning\nparams:\n"
    << "  - name: eps_start\n    type: float\n    default: " << eps_start
    << "\n    min: 1.0\n    max: 10.0\n    description: start\n"
    << "  - name: eps_final\n    type: float\n    default: " << eps_final
    << "\n    min: 1.0\n    max: 10.0\n    description: final\n"
    << "  - name: eps_step\n    type: float\n    default: " << eps_step
    << "\n    min: 0.01\n    max: 10.0\n    description: step\n"
    << "  - name: max_expansions\n    type: int\n    default: 100000\n    min: 1\n"
    << "    max: 100000000\n    description: cap\n";
  return core::ParamSet::from_yaml(test::write_temp("ara_star.yaml", y.str()));
}

maps::OccupancyGrid2D& as_grid(core::MapBase& m) {
  return *dynamic_cast<maps::OccupancyGrid2D*>(&m);
}

}  // namespace

// (a) known map: ARA*'s final (eps -> 1.0) solution equals the A* optimum ---------
TEST(AraStar, ConvergesToAstarOptimumOnMaze01) {
  auto map = maps::load_map(test::repo_path("maps/grid/maze01.yaml"), 0, 8);
  auto& grid = as_grid(*map);
  auto sc = maps::load_scenario(test::repo_path("maps/scenarios/maze01_s1.yaml"));
  Cell start = grid.world_to_cell(sc.start.x, sc.start.y);
  Cell goal = grid.world_to_cell(sc.goal.x, sc.goal.y);

  global_planning::AraStar ara(cfg("ara_star"));
  auto ra = ara.plan(grid, start, goal, nullptr);
  global_planning::AstarPlanner astar(cfg("astar"));
  auto ro = astar.plan(grid, start, goal, nullptr);

  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_NEAR(ra.cost, ro.cost, 1e-9);  // final ARA* == A* optimum
}

// (b) no-path case ----------------------------------------------------------------
TEST(AraStar, NoPathWhenWalledOff) {
  auto g = test::make_grid({"..#..", "..#..", "..#..", "..#..", "..#.."});
  global_planning::AraStar ara(cfg("ara_star"));
  auto r = ara.plan(g, Cell{0, 0}, Cell{0, 4}, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  EXPECT_EQ(r.cost, 0.0);
}

// (c) param validation failure ----------------------------------------------------
TEST(AraStar, BadEpsStartThrows) {
  std::string bad = test::write_temp(
      "ara_star.yaml",
      "algorithm: ara_star\ncategory: global_planning\nparams:\n"
      "  - name: eps_start\n    type: float\n    default: 0.5\n    min: 1.0\n    max: 10.0\n"
      "    description: below min\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

// (d) anytime property: an inflated first solution, then repair to the optimum.
// wastar_greedy01 has two obstacles an inflated heuristic rounds on the costly side
// first, so the eps=3 solution is strictly worse than the eps->1 optimum.
TEST(AraStar, AnytimeSolutionsImproveToOptimum) {
  auto map = maps::load_map(test::repo_path("maps/grid/wastar_greedy01.yaml"), 0, 8);
  auto& grid = as_grid(*map);
  auto sc = maps::load_scenario(test::repo_path("maps/scenarios/wastar_greedy01_s1.yaml"));
  Cell start = grid.world_to_cell(sc.start.x, sc.start.y);
  Cell goal = grid.world_to_cell(sc.goal.x, sc.goal.y);

  // First solution only: a single weighted-A* pass at eps=3 (eps_final=3).
  global_planning::AraStar first(ara_cfg(3.0, 3.0, 1.0));
  auto rf = first.plan(grid, start, goal, nullptr);
  ASSERT_TRUE(rf.success);

  // Full sweep eps 3 -> 1, capturing the emitted anytime solutions.
  std::ostringstream os;
  core::TraceRecorder rec(os);
  global_planning::AraStar ara(ara_cfg(3.0, 1.0, 1.0));
  auto ra = ara.plan(grid, start, goal, &rec);
  ASSERT_TRUE(ra.success);

  int path_found = 0;
  std::istringstream in(os.str());
  for (std::string line; std::getline(in, line);) {
    if (line.find("\"event\":\"path_found\"") != std::string::npos) ++path_found;
  }
  EXPECT_GE(path_found, 2);  // genuinely anytime: at least one repair after the first

  global_planning::AstarPlanner astar(cfg("astar"));
  auto ro = astar.plan(grid, start, goal, nullptr);
  EXPECT_NEAR(ra.cost, ro.cost, 1e-9);  // converged to the optimum
  EXPECT_GT(rf.cost, ra.cost);          // first eps-inflated path is strictly suboptimal
}

File: python/tests/test_ara_star.py
"""ARA*: converges to the A* optimum (eps -> 1), the no-path case, param
validation, and the defining anytime property (each solution no worse than the
next, first >= last)."""

from __future__ import annotations

import io
import json
from pathlib import Path

import pytest
from conftest import REPO_ROOT, config, grid_from, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell
from navigation.global_planning import AStar
from navigation.global_planning.search import ARAStar
from navigation.maps.loader import load_map, load_scenario

# wastar_greedy01: an open field with two obstacles that lie between start and goal,
# crafted so an inflated heuristic rounds them on the costly side first (see the A*
# counterexample). With eps 3 -> 1 ARA* returns a suboptimal path, then repairs to
# the optimum — an unambiguous anytime demonstration.
_WASTAR_ROWS = [
    ".............",
    "..........#..",
    "...........#.",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
]
_WASTAR_START: Cell = (11, 12)
_WASTAR_GOAL: Cell = (0, 11)


def _path_cost(grid: object, path: list[Cell]) -> float:
    total = 0.0
    for a, b in zip(path, path[1:], strict=False):
        for cell, cost in grid.neighbors(a):  # type: ignore[attr-defined]
            if cell == b:
                total += cost
                break
    return total


def _anytime_config(tmp_path: Path, eps_start: float, eps_step: float) -> ParamSet:
    return ParamSet.from_yaml(
        write_config(
            tmp_path / "ara_any.yaml",
            "ara_star",
            [
                {"name": "eps_start", "type": "float", "default": eps_start,
                 "min": 1.0, "max": 10.0, "description": "start eps"},
                {"name": "eps_final", "type": "float", "default": 1.0,
                 "min": 1.0, "max": 10.0, "description": "final eps"},
                {"name": "eps_step", "type": "float", "default": eps_step,
                 "min": 0.01, "max": 10.0, "description": "eps decrement"},
                {"name": "max_expansions", "type": "int", "default": 100000,
                 "min": 1, "max": 100000000, "description": "cap"},
            ],
        )
    )


def test_ara_star_converges_to_astar_optimum_on_maze01() -> None:
    # eps -> eps_final (1.0) makes ARA*'s final solution the admissible A* optimum.
    grid = load_map(str(REPO_ROOT / "maps/grid/maze01.yaml"), seed=0, connectivity=8)
    sc = load_scenario(str(REPO_ROOT / "maps/scenarios/maze01_s1.yaml"))
    start = grid.world_to_cell(*sc.start)
    goal = grid.world_to_cell(*sc.goal)
    ara = ARAStar(config("ara_star")).plan(grid, start, goal)
    ast = AStar(config("astar")).plan(grid, start, goal)
    assert ara.success and ast.success
    assert ara.path[0] == start and ara.path[-1] == goal
    assert ara.cost == pytest.approx(ast.cost)  # final ARA* == A* optimum


def test_ara_star_no_path_when_walled_off() -> None:
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."])
    res = ARAStar(config("ara_star")).plan(grid, (0, 0), (0, 4))
    assert not res.success
    assert res.path == []
    assert res.cost == 0.0


def test_ara_star_rejects_out_of_range_param(tmp_path: Path) -> None:
    # eps_start declares min 1.0; a below-min default must fail at load time so an
    # inflation factor < 1 (which would break the suboptimality bound) never runs.
    cfg = write_config(
        tmp_path / "bad_ara.yaml",
        "ara_star",
        [{"name": "eps_start", "type": "float", "default": 0.5,
          "min": 1.0, "max": 10.0, "description": "below min"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(cfg)


def test_ara_star_anytime_solutions_improve_monotonically(tmp_path: Path) -> None:
    grid = grid_from(_WASTAR_ROWS)
    cfg = _anytime_config(tmp_path, eps_start=3.0, eps_step=1.0)
    buf = io.StringIO()
    res = ARAStar(cfg).plan(grid, _WASTAR_START, _WASTAR_GOAL, TraceRecorder(buf))

    events = [json.loads(line) for line in buf.getvalue().splitlines()]
    solutions = [[tuple(s) for s in e["path"]] for e in events if e["event"] == "path_found"]
    costs = [_path_cost(grid, p) for p in solutions]

    assert len(costs) >= 2  # genuinely anytime: at least one repair after the first
    assert costs == sorted(costs, reverse=True)  # each solution no worse than the next
    assert costs[0] > costs[-1]  # the first eps-inflated path is strictly suboptimal here
    ast = AStar(config("astar")).plan(grid, _WASTAR_START, _WASTAR_GOAL)
    assert res.cost == pytest.approx(costs[-1])
    assert res.cost == pytest.approx(ast.cost)  # converged to the optimum
